server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -156,7 +157,7 @@ public void clusterChanged(ClusterChangedEvent event) {
                 synchronized (shardSnapshots) {
                     cancelRemoved(currentSnapshots);
                     if (currentSnapshots != null) {
-                        startNewSnapshots(currentSnapshots);
+                        startNewSnapshots(currentSnapshots, event.state().nodes().getMinNodeVersion());
```

Comment:
I think it's dangerous if the node decide whether to use old-style or new-style all by themselves as they sample the cluster state at different points in time, and therefore making different decisions. I would prefer for the master to establish this at the beginning of taking the snapshot, and then put "old-style" or "new-style" into the SnapshotInProgress.Entry of the cluster state, and all nodes just execute on that.

Reply:
I figured it wouldn't matter since we never downgrade? Old-style is understood by all nodes and in case we only have new-style nodes in the CS at one point, that will never change will it?

But regardless, we can obviously track this in the CS if you want. Maybe we could just add a `version` field to `SnapshotInProgress.Entry`, then we can use it for follow-up changes BwC as well?

Reply:
Update: I thought about this some more and I don't think it adds anything to track the style in the `SnapshotsInProgress.Entry` relative to the current solution.
The current solution can't move from a new version cluster back to a mixed version cluster which would be the only problematic case. In this case (or a failover from a new version to an old version master) the information about the style of snapshot would be lost anyway since the old version master can't track that new field. Obviously you would default it to old-style then, but it seems completely redundant and just adds more complication to `SnapshotsInProgress` serialization to no benefit?

Reply:
> The current solution can't move from a new version cluster back to a mixed version cluster

That's true. But for a particular snapshot, the transition can happen mid-way through the snapshot. That means that some data nodes write old-style, others new-style, and same goes for the master. I wonder if that will make things more complex than just sticking to one consistent style throughout a given snapshot.

Reply:
> That means that some data nodes write old-style, others new-style, and same goes for the master. 

That's true, but note that creating a snapshot only has two steps on the repository level these days (we had three steps before since we wrote the clusterstate + index-metadata first but now we do that during finalization).
So even if data nodes write a mix, they will also send their new generation to `master` and `master` will finalize in the new format with some gens as numeric values instead of uuids.
The only maybe-issue I see is that we're currently not deleting the last numeric-suffix index-N on snapshot create (I could add that back, or we could ignore it since those get cleaned up during the next delete/cleanup anyway and don't take much space in the first place ... might not be worth the added code complexity).

Reply:
Update:
The above was slighly stupid. I added back the BwC deletion of old index-N from primaries since obviously the mixed-cluster state could run for multiple snapshots and we'd continue to leak index-N in that case without that. I can't see any BwC case that isn't neatly handled now though.
Even if all data nodes were to write in old style and the master would finalize in new-style it'd be fine. Whether the master was in new or old-style mode when the whole snapshot started is entirely irrelevant to the snapshot process sine the master doesn't do any writes on snapshot initialization now.
Since it's impossible for a data-node to write a new-style index-${uuid} blob and the master to not finalize in the new-style I don't think any problems are possible here still.

Reply:
> If put in SnapshotsInProgress, I would prefer a simple boolean. In master, that code can go away anyway.

`boolean` it is then :)

> I also wonder if by using null in ShardGenerations we can't distinguish between "there is no metadata file that we care about for this shard" and "the snapshot of this shard was taken while the cluster was on a version prior to 7.5.0".

This is a fair point. The current logic of trying to find a numeric `index-N` doesn't make sense in case of a newly created shared or index path. I'll put a placeholder for that case we can avoid the redundant listing of empty/new shard paths.

Reply:
Added placeholder logic for known new shards in b9dbd7c5471 saving us a `LIST` for new indices (the removed BwC path in `RepositoryData` is irrelevant nowadays anyway).